src/mlpack/methods/det/dtree_impl.hpp

```diff
@@ -197,44 +308,47 @@ bool DTree::FindSplit(const arma::mat& data,
     double minDimError = std::pow(points, 2.0) / (max - min);
     double dimLeftError = 0.0; // For -Wuninitialized.  These variables will
     double dimRightError = 0.0; // always be set to something else before use.
-    double dimSplitValue = 0.0;
+    ElemType dimSplitValue = 0.0;
 
     // Find the log volume of all the other dimensions.
     double volumeWithoutDim = logVolume - std::log(max - min);
 
-    // Get the values for the dimension.
-    arma::rowvec dimVec = data.row(dim).subvec(start, end - 1);
-
-    // Sort the values in ascending order.
-    dimVec = arma::sort(dimVec);
-
-    // Find the best split for this dimension.  We need to figure out why
-    // there are spikes if this minLeafSize is enforced here...
-    for (size_t i = minLeafSize - 1; i < dimVec.n_elem - minLeafSize; ++i)
+    // Get the values for splitting. The old implementation:
+    //   dimVec = data.row(dim).subvec(start, end - 1);
+    //   dimVec = arma::sort(dimVec);
+    // could be quite inefficient for sparse matrices, due to copy operations (3).
+    // This one has custom implementation for dense and sparse matrices.
+
+    std::vector<SplitItem> splitVec = details::ExtractSplits(data,
+                                                             dim,
+                                                             start,
+                                                             end,
+                                                             minLeafSize);
```

Comment:
As far as I can tell the reason for the `ExtractSplits` function is because the `sort()` method is not available for sparse matrices.  Suppose that `sort()` did exist for sparse matrices (e.g. suppose I sat down and wrote it, which I might need to do shortly!).  Then we could do this...

```
typename MatType::row_type dimVec = data.row(dim).subvec(start, end - 1);
dimVec = arma::sort(dimVec);

// Iterate over all possible values.
typename MatType::row_type::const_row_col_iterator it;
for (it = dimVec.begin_row_col(); ++it; it != dimVec.end_row_col())
{
  // Check the split to the left side of the point that *it represents, if it exists.
  if (it->col() > 0)
  {
    // do checking for split between dimVec[it->col() - 1] and dimVec[it->col()]...
  }

  // If we are in the next-to-last position, check the split to the right, if applicable.
  // There's probably a cleaner way to write this code.
  typename MatType::row_type::const_row_col_iterator it2 = it;
  if ((++it2) == dimVec.end_row_col())
  {
    // do checking for split between dimVec[it->col()] and dimVec[it->col() + 1]...
  }
}
```

Note that the `row_col_iterator` will only "stop" at points that are actually represented in memory.  So for sparse matrices it will skip over zero elements.  I think that the `row_col_iterator` is not actually documented by Armadillo... I should submit a patch for that...

What do you think?  Would this approach work?  If so I will implement `SpMat::sort()` (it should be pretty straightforward, I think I have a good idea).  That would allow us to avoid having specific code for both the dense and sparse case (I like to push specific code like that to Armadillo wherever possible).